### DISCUSSION TITLE
`[` and `]` should have a modifier of typeHint when used in type hints

### DISCUSSION THREAD

**Discussion by NorthIsUp** [microsoft/pylance-release#2680] [Enhancement]
## Environment data

Name: Pylance
Id: ms-python.vscode-pylance
Description: A performant, feature-rich language server for Python in VS Code
Version: 2021.10.2-pre.1
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-python.vscode-pylance

-   Language Server version: ^^
-   OS and version: ubuntu 18
-   Python version (& distribution if applicable, e.g. Anaconda): 3.7.6

## Expected behaviour

Semantic coloring should be consistent in the type hint, but the item access `[` and `]` don't get the `typeHint` modifier

## Actual behaviour

The item accessors `[` and `]` don't get the `typeHint` modifier in a type hint

## Code Snippet / Additional information

```python
from typing import List

# I expect the "[" to have a modifier of "typeHint"
def foo(some_stuff: List[str]) -> None:
    return ['hi']
```

in this screenshot i'd expect the `[` and `]` to be red like the rest of the type hits.
![CleanShot 2021-10-20 at 15 23 49](https://user-images.githubusercontent.com/122612/138158486-46bf4826-2e4e-4b81-921b-25dfbea3676b.png)

**Comment by jakebailey** [comment]
We don't emit semantic tokens for any tokens like this (only names, builtin constants `None`/`True`/`False`/`__debug__`, and `match`/`case`). I'm not certain it's by design (similarly reported in #459 with other bugs), but our parse tree doesn't always have the information to make it straightforward to color these types of tokens, and we've relied on VS Code's TextMate profile to do the rest.

Reading the LSP spec for semantic tokens, I'm not entirely certain there's a token that we can use for these, though: https://microsoft.github.io/language-server-protocol/specification#textDocument_semanticTokens

We'd have to compare to something like TS with its `<>` or similar.

**Comment by yvvt0379** [comment]
From my perspective, I think this can be an option(not as default), because when a type hint is complicated(like `dict[int, Union[str, Iterable[str]]]`), highlighting square brackets the same color as types just makes the code harder to read & understand.

**Comment by NorthIsUp** [comment]
> From my perspective, I think this can be an option(not as default), because when a type hint is complicated(like `dict[int, Union[str, Iterable[str]]]`), highlighting square brackets the same color as types just makes the code harder to read & understand.

agreed, ideally they could be colored as their own entity like `typehint.item-access` or something.

**Comment by heejaechang** [comment]
implementation wise, we will need to go through tokens to support this since for now, we only process parse tree and name node.

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.
